core/src/transport/map_err.rs

```diff
@@ -76,73 +77,49 @@ where
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
         let map = self.map.clone();
         match self.transport.dial_as_listener(addr) {
-            Ok(future) => Ok(MapErrDial {
+            Ok(future) => Ok(MapErrFuture {
                 inner: future,
                 map: Some(map),
+                _marker: PhantomData,
             }),
             Err(err) => Err(err.map(map)),
         }
     }
 
-    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
-        self.transport.address_translation(server, observed)
-    }
-
     fn poll(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
         let this = self.project();
         let map = &*this.map;
         this.transport.poll(cx).map(|ev| {
-            ev.map_upgrade(move |value| MapErrListenerUpgrade {
+            ev.map_upgrade(move |value| MapErrFuture {
                 inner: value,
                 map: Some(map.clone()),
+                _marker: PhantomData,
             })
             .map_err(map.clone())
         })
     }
-}
-
-/// Listening upgrade future for `MapErr`.
-#[pin_project::pin_project]
-pub struct MapErrListenerUpgrade<T: Transport, F> {
-    #[pin]
-    inner: T::ListenerUpgrade,
-    map: Option<F>,
-}
-
-impl<T, F, TErr> Future for MapErrListenerUpgrade<T, F>
-where
-    T: Transport,
-    F: FnOnce(T::Error) -> TErr,
-{
-    type Output = Result<T::Output, TErr>;
 
-    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let this = self.project();
-        match Future::poll(this.inner, cx) {
-            Poll::Ready(Ok(value)) => Poll::Ready(Ok(value)),
-            Poll::Pending => Poll::Pending,
-            Poll::Ready(Err(err)) => {
-                let map = this.map.take().expect("poll() called again after error");
-                Poll::Ready(Err(map(err)))
-            }
-        }
+    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.transport.address_translation(server, observed)
     }
 }
 
-/// Dialing future for `MapErr`.
+/// Future for `MapErr`.
 #[pin_project::pin_project]
-pub struct MapErrDial<T: Transport, F> {
+pub struct MapErrFuture<T, Fut, F> {
```

Comment:
This is also breaking. Please add a type alias to make it non-breaking with a deprecation attribute.